swap/swap.go

```diff
@@ -17,69 +17,340 @@
+	privateKey *ecdsa.PrivateKey // private key
+	publicKey  *ecdsa.PublicKey  // public key
+}
+
+// Params encapsulates param
+type Params struct {
+	InitialDepositAmount uint64 //
+}
+
+// NewDefaultParams returns a Params struct filled with default values
+func NewDefaultParams() *Params {
+	return &Params{
+		InitialDepositAmount: DefaultInitialDepositAmount,
+	}
 }
 
 // New - swap constructor
-func New(stateStore state.Store) (swap *Swap) {
-	swap = &Swap{
-		stateStore: stateStore,
-		balances:   make(map[enode.ID]int64),
+func New(stateStore state.Store, prvkey *ecdsa.PrivateKey, contract common.Address, backend cswap.Backend) *Swap {
+	sw := &Swap{
+		stateStore:          stateStore,
+		balances:            make(map[enode.ID]int64),
+		backend:             backend,
+		cheques:             make(map[enode.ID]*Cheque),
+		peers:               make(map[enode.ID]*Peer),
+		params:              NewDefaultParams(),
+		paymentThreshold:    DefaultPaymentThreshold,
+		disconnectThreshold: DefaultDisconnectThreshold,
+		contractReference:   nil,
+		oracle:              NewPriceOracle(),
+	}
+	sw.owner = sw.createOwner(prvkey, contract)
+	return sw
+}
+
+// createOwner assings keys and addresses
+func (s *Swap) createOwner(prvkey *ecdsa.PrivateKey, contract common.Address) *Owner {
+	pubkey := &prvkey.PublicKey
+	return &Owner{
+		privateKey: prvkey,
+		publicKey:  pubkey,
+		Contract:   contract,
+		address:    crypto.PubkeyToAddress(*pubkey),
 	}
-	return
 }
 
-//Swap implements the protocols.Balance interface
-//Add is the (sole) accounting function
+// DeploySuccess is for convenience log output
+func (s *Swap) DeploySuccess() string {
+	return fmt.Sprintf("contract: %s, owner: %s, deposit: %v, signer: %x", s.owner.Contract.Hex(), s.owner.address.Hex(), s.params.InitialDepositAmount, s.owner.publicKey)
+}
+
+// Add is the (sole) accounting function
+// Swap implements the protocols.Balance interface
 func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	//load existing balances from the state store
+	// load existing balances from the state store
 	err = s.loadState(peer)
 	if err != nil && err != state.ErrNotFound {
+		log.Error("error while loading balance for peer", "peer", peer.ID().String())
 		return
 	}
+
+	// Check if balance with peer is over the disconnect threshold
+	// It is the creditor who triggers the disconnect from a overdraft creditor,
+	// thus we check for a positive value
+	if s.balances[peer.ID()] >= s.disconnectThreshold {
+		// if so, return error in order to abort the transfer
+		disconnectMessage := fmt.Sprintf("balance for peer %s is over the disconnect threshold %v, disconnecting", peer.ID().String(), s.disconnectThreshold)
+		log.Warn(disconnectMessage)
+		return errors.New(disconnectMessage)
+	}
+
+	// calculate new balance
+	var newBalance int64
+	newBalance, err = s.updateBalance(peer.ID(), amount)
+	if err != nil {
+		return
```

Comment:
return err

Reply:
what about a possible rollback of the balance? are you keeping this in mind?

Reply:
Any specific issue in mind?

This is a can of worms and very difficult to tackle in a p2p accounting IMHO. Although we *are trying* keeping this in mind where we think it makes sense, we would then have to address things like:

* what happens if node A emits a cheque, at which point it *already resets the balance with the peer*, but the peer never received the cheque message because it went down?

In any case we have been clear that this PR is actually some "happy-day scenario" (@mortelli please add to the PR description in a meaningful way), and that better error handling will be needed in further iterations

Reply:
Edited the PR description with additional **bold text** for further clarification.

Reply:
The assumption MUST be that if `updateBalance` failed then the balance was not updated and thus there is nothing to roll back